Create QML test automation target (qml/test/test_bitcoin-qt)

johnny9 opened the pull request:
Adds src/qml/test/test_bitcoin-qt as a new test target. This program is built using the QuickTest library. It can be used to validate our qml code by rederimg our components and providing simulated user input. It is set to only be built if QuickTest library is found, gui tests are enabled, and bitcoin-qt is configured for qml.

As a part of test setup, a test implementation of our imageprovider is added to the test target to avoid bringing in too many dependencies. Additionally, a fake AppMode component is added to mock setting the device layout mode. 

More information about the QuickTest library can be found at https://doc.qt.io/qt-5.15/qtquicktest-index.html.


[![Windows](https://img.shields.io/badge/OS-Windows-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/215)
[![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/215)
[![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/215)
[![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/insecure_android_apk.zip?branch=pull/215)

Commit:
qml: add makefile for qmltest target

Commit:
qml: add test harness for onboarding qml tests

Commit:
qml: add QtQuickTest library to qmltest target

Commit:
qml: add first qml testcase

Commit:
qml: add imageprovider to test setup

Commit:
qml: remove double linkage in Makefile.qmltest.include

Commit:
qml/test: have OnboardingCover fill parent

Commit:
qml/test: add fake qt.bitcoincore.org module

Commit:
qml: extract OnboardingWizard from main.qml

Commit:
qml/test: add test_all_continue_buttons

Commit:
qmltest: configure to only compile with qml and tests enabled

Commit:
qmltest: rename testimageprovider

jarolrod reviewed (commented):
Concept ACK, a good start. Testing that we can progress through onboarding is a great first test case. Will look into failures.

Commit:
qml/test: add include guards to headers

Commit:
gitignore: add output from qmltest

Commit:
qmltest: don't compile if qml isn't enabled

hebasto commented:
@johnny9 

Thank you very much for working on tests!

Mind adding some words to the PR description?

hebasto commented on src/qml/test/tst_onboarding.qml:
```diff
@@ -0,0 +1,60 @@
+
+        function test_all_continue_buttons() {
+            verify(onboardingWizard.currentIndex == 0)
+ 
```
Needed to make `./test/lint/lint-whitespace.py` happy:
```suggestion

```

In general, suggesting ensure that linters are happy for each commit.

hebasto commented on src/Makefile.qmltest.include:
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2013-2023 The Bitcoin Core developers
```
```suggestion
# Copyright (c) 2023 The Bitcoin Core developers
```

hebasto commented on src/qml/test/onboardingtests.h:
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2014-2023 The Bitcoin Core developers
```
```suggestion
// Copyright (c) 2023 The Bitcoin Core developers
```

hebasto commented on src/qml/test/org/bitcoincore/qt/AppMode.qml:
```diff
@@ -0,0 +1,12 @@
+pragma Singleton
```
Add a copyright header?

hebasto reviewed (commented):
Concept ACK.

If there are no reasons against it, maybe rename s/`tst_onboarding.qml`/`test_onboarding.qml`/?

johnny9 commented:
tst_*.qml is actually the pattern the library uses to find test cases in a directory.

johnny9 commented:
Still have some linking issues with Makefile.qmltest.include it seems as CI isn't happy. I likely linked too make things into the binary.

hebasto commented:
> Still have some linking issues with Makefile.qmltest.include it seems as CI isn't happy. I likely linked too make things into the binary.

Looking into them...

hebasto commented on src/qml/bitcoin_qml.qrc:
```diff
@@ -34,6 +34,7 @@
         <file>pages/onboarding/OnboardingStorageAmount.qml</file>
         <file>pages/onboarding/OnboardingStorageLocation.qml</file>
         <file>pages/onboarding/OnboardingStrengthen.qml</file>
+        <file>pages/onboarding/OnboardingWizard.qml</file>
```
We don't have `pages/onboarding/OnboardingWizard.qml`, do we?

johnny9 replied:
Good catch, missed a git add. It is included in the PR now.

hebasto commented:
Static builds, which all builds with depends are, require importing of plugins like this:
```diff
--- a/src/qml/test/onboardingtests.cpp
+++ b/src/qml/test/onboardingtests.cpp
@@ -1,11 +1,32 @@
 // Copyright (c) 2023 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <qml/test/onboardingtests.h>
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
 #include <QtQuickTest>
 #include <QQmlEngine>
 #include <QQmlContext>
 #include <qml/test/testimageprovider.h>
-#include <qml/test/onboardingtests.h>
+
+#if defined(QT_STATICPLUGIN)
+#include <QtPlugin>
+Q_IMPORT_PLUGIN(QtQuick2Plugin)
+Q_IMPORT_PLUGIN(QTestQmlModule)
+#if defined(QT_QPA_PLATFORM_XCB)
+Q_IMPORT_PLUGIN(QXcbIntegrationPlugin);
+#elif defined(QT_QPA_PLATFORM_WINDOWS)
+Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin);
+#elif defined(QT_QPA_PLATFORM_COCOA)
+Q_IMPORT_PLUGIN(QCocoaIntegrationPlugin);
+#elif defined(QT_QPA_PLATFORM_ANDROID)
+Q_IMPORT_PLUGIN(QAndroidPlatformIntegrationPlugin)
+#endif
+#endif
 
 Setup::Setup()
 {
```

And add to `build-aux/m4/bitcoin_qt.m4` a few lines `QT_LIBS="$QT_LIBS -L$qt_plugin_path/../qml/QtTest"` and `_BITCOIN_QT_CHECK_STATIC_PLUGIN([QTestQmlModule], [-lqmltestplugin])` (did not test yet).

hebasto commented:
Maybe, label this PR as a draft for now?

hebasto commented on build-aux/m4/bitcoin_qt.m4:
```diff
@@ -467,6 +471,10 @@ AC_DEFUN([_BITCOIN_QT_FIND_LIBS],[
                         [BITCOIN_QT_FAIL([${qt_lib_prefix}QuickControls2${qt_lib_suffix} $qt_version not found])])
     ])
+    BITCOIN_QT_CHECK([
+      PKG_CHECK_MODULES([QT_QUICKTEST], [${qt_lib_prefix}QuickTest${qt_lib_suffix} $qt_version], [QT_QMLTEST_INCLUDES="$QT_QUICKTEST_CFLAGS"; have_quick_test=yes],[have_quick_test=no])
```
Introducing `QT_QMLTEST_INCLUDES` requires `AC_SUBST(QT_QMLTEST_INCLUDES)` to be able to use it `Makefile`s.

hebasto commented:
FWIW, I'm testing this PR on Ubuntu 22.04 with depends:
```
$ make -C depends NO_WALLET=1 NO_ZMQ=1 NO_UPNP=1 NO_NATPMP=1 NO_QR=1
$ ./autogen.sh
$ ./configure --disable-fuzz-binary --disable-bench CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site
$ make clean
$ make
```

hebasto commented:
From https://doc.qt.io/qt-5/qtquicktest-index.html:
> Note: There is no binary compatibility guarantee for the Qt Quick Test module. This means that an application that uses Qt Quick Test is only guaranteed to work with the Qt version it was developed against. However, source compatibility is guaranteed.

Should we skip Quick test when building with depends?

johnny9 commented:
Thank you for the review Hebasto. Very helpful. Marked as a draft and will unmark after addressing the issues you've mentioned and am more confident it can get through CI.

Commit:
qmltest: rename onboardingtests to setup

Commit:
qmltest: add/fix copyright comments

Commit:
qmltest: remove extra flags/link aguments from makefile

hebasto commented:
@johnny9 

[Here](https://github.com/hebasto/gui-qml/commits/pr215/alt) is a quick and dirty branch which is built [successfully](https://cirrus-ci.com/build/5945618690998272) in all CI tasks. However, running `qml/test/test_bitcoin-qt` in the CI environment needs more investigation regarding which value of the `QT_QPA_PLATFORM` variable should be used.

Feel free to use my branch at your discretion.

johnny9 commented:
> @johnny9 
> 
> [Here](https://github.com/hebasto/gui-qml/commits/pr215/alt) is a quick and dirty branch which is built [successfully](https://cirrus-ci.com/build/5945618690998272) in all CI tasks. However, running `qml/test/test_bitcoin-qt` in the CI environment needs more investigation regarding which value of the `QT_QPA_PLATFORM` variable should be used.
> 
> Feel free to use my branch at your discretion.

Thank you. I'll take a look at it and merge it with my branch. I made an attempt last night but struggled with the QT STATIC PLUGIN checks. I'll be investigating how to get this supported in cirrus afterwards. I have some ideas but it'll depend on what kind of platform support they have for a buffer to render into. I'd like to use this to validate the qml in our PRs and maybe even generate screenshots since that eats up a lot of review time.

hebasto commented:
> I made an attempt last night but struggled with the QT STATIC PLUGIN checks.

Still thinking it doesn't worth. See https://github.com/bitcoin-core/gui-qml/pull/215#issuecomment-1382813668.

Commit:
qmltest: fix build and don't build when static linking

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

hebasto commented:
The `qml/pages/onboarding/OnboardingWizard.qml ` is still to be added to `QML_RES_QML`.

Commit:
qmltest: remove from TESTS until able to run in ci

hebasto commented:
Might the introducing of `src/qml/pages/onboarding/OnboardingWizard.qml` be split up into its own PR?

johnny9 commented:
> Might the introducing of `src/qml/pages/onboarding/OnboardingWizard.qml` be split up into its own PR?

Yeah I can do that. It was done to simplify testing but the change is generally good as the main,qml gets simpler.

johnny9 commented:
Update from[ 75da2a2 ](https://github.com/bitcoin-core/gui-qml/commit/75da2a2863f3b665cda34b8be9631f0a0a462549)to[ ed8222b](https://github.com/bitcoin-core/gui-qml/commit/ed8222b7dfb87d6150ad17cd71bfed76d0745202)
 - Rebased with main

jarolrod reviewed (commented):
General Concept ACK

Test cases work and run
```
********* Start testing of onboarding *********
Config: Using QtTest library 5.15.10, Qt 5.15.10 (arm64-little_endian-lp64 shared (dynamic) release build; by Clang 14.0.0 (clang-1400.0.29.202) (Apple)), osx 13.5
PASS   : onboarding::initTestCase()
PASS   : onboarding::test_all_continue_buttons()
PASS   : onboarding::cleanupTestCase()
Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 1042ms
********* Finished testing of onboarding *********
```

jarolrod commented on src/qml/test/testimageprovider.cpp:
```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2023 The Bitcoin Core developers
```
There has to be a way for us to not have to copy over this into qml/test

jarolrod commented on src/qml/test/tst_onboarding.qml:
```diff
@@ -0,0 +1,64 @@
+
+            var cover = findChild(onboardingWizard, "onboardingCover")
+            verify(cover, onboardingWizard.currentItem)
+            mouseClick(findChild(cover, "continueButton"))
```
What about abstracting a lot of the repetitive action of finding and clicking into a helper function?

helper function looking something like:
```
function clickContinueAndVerify(pageId, expectedIndex) {
    var page = findChild(onboardingWizard, pageId)
    verify(page, onboardingWizard.currentItem)
    mouseClick(findChild(page, "continueButton"))
    verify(onboardingWizard.currentIndex == expectedIndex)
}
```

Then calling like:
```
clickContinueAndVerify("onboardingCover", 1)
wait(200)
clickContinueAndVerify("onboardingStrengthen", 2)
```

jarolrod replied:
A lot of tests would involve a repetitive function, so good to think about this now to establish a style for how we write our tests

jarolrod commented:
Need to cleanup the commit history :D

pablomartin4btc reviewed (commented):
Concept ACK

It would be good to have this soon.

tested ed8222b7dfb87d6150ad17cd71bfed76d0745202 on Ubuntu 22.04.

```
********* Start testing of onboarding *********
Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 11.3.0), ubuntu 22.04
PASS   : onboarding::initTestCase()
PASS   : onboarding::test_all_continue_buttons()
PASS   : onboarding::cleanupTestCase()
Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 1049ms
********* Finished testing of onboarding *********
```

Apart from @jarolrod's comments, since we are moving away from `SwipeView` I think we'd need to update this to `PageStacks` (#428).

At least in Ubuntu, I see the UI gets opened and pages are switching, not sure if this is on purpose and can be hidden (I haven't checked the code) as in current Qt this is not happening.

<details>
<summary>I got these errors before the successful results pasted above.</summary>

```
/src/qml/test/test_bitcoin-qt 
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/controls/Theme.qml:13:5: QML Settings: Failed to initialize QSettings instance. Status code is: 1
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/controls/Theme.qml:13:5: QML Settings: The following application identifiers have not been set: QVector("organizationName", "organizationDomain")
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageOptions.qml:33: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/AboutOptions.qml:54: ReferenceError: nodeModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/DeveloperOptions.qml:20: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/DeveloperOptions.qml:45: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/DeveloperOptions.qml:16: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/DeveloperOptions.qml:41: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageOptions.qml:27: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageOptions.qml:25: ReferenceError: chainModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageOptions.qml:42: ReferenceError: chainModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageOptions.qml:41: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageSettings.qml:19: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/StorageSettings.qml:43: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/ConnectionSettings.qml:17: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/ConnectionSettings.qml:30: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/ConnectionSettings.qml:43: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/components/ConnectionSettings.qml:56: ReferenceError: optionsModel is not defined
file:///home/pablo/workspace/bitcoin-core-gui-qml/src/qml/pages/onboarding/OnboardingStorageLocation.qml:23: ReferenceError: chainModel is not defined
********* Start testing of onboarding *********
Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 11.3.0), ubuntu 22.04
PASS   : onboarding::initTestCase()
PASS   : onboarding::test_all_continue_buttons()
PASS   : onboarding::cleanupTestCase()
Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 1049ms
********* Finished testing of onboarding *********

```

</details>

pablomartin4btc commented:
During discussions with @jarolrod about reorganizing the project structure (components, pages, controls, etc.), we noted that tests will need to be updated once Qt 6 is introduced, along with CMake. Considering this, it might be better to treat this as a draft for now, until we decide whether to create a new PR from scratch.

hebasto commented:
Closing for now.

Feel free to reopen this based on the new [default](https://github.com/bitcoin-core/gui-qml/tree/qt6) branch.
